source/sigv4.c

```diff
@@ -1162,110 +1620,1287 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+
+/*-----------------------------------------------------------*/
+
+    static void setQueryStringFieldsAndValues( const char * pQuery,
+                                               size_t queryLen,
+                                               size_t * pNumberOfParameters,
+                                               CanonicalContext_t * pCanonicalRequest )
     {
-        LogError( ( "Parameter check failed: dateLen must be either %u or %u, "
-                    "for RFC 3339 and RFC 5322 formats, respectively.",
-                    SIGV4_EXPECTED_LEN_RFC_3339,
-                    SIGV4_EXPECTED_LEN_RFC_5322 ) );
+        size_t currentParameter = 0U, i = 0U, startOfFieldOrValue = 0U;
+        uint8_t fieldHasValue = 0U;
+
+        /* Note: Constness of the query string is casted out here, taking care not to modify
+         * its contents in any way. */
+
+        /* Set cursors to each field and value in the query string. */
+        for( i = 0U; i < queryLen; i++ )
+        {
+            if( ( pQuery[ i ] == '=' ) && !fieldHasValue )
+            {
+                setQueryParameterKey( currentParameter, &pQuery[ startOfFieldOrValue ], i - startOfFieldOrValue, pCanonicalRequest );
+                startOfFieldOrValue = i + 1U;
+                fieldHasValue = 1U;
+            }
+            else if( ( i == queryLen - 1U ) || ( ( pQuery[ i ] == '&' ) && ( i != 0U ) ) )
+            {
+                /* Adjust for the length of the last query parameter. */
+                if( i == queryLen - 1U )
+                {
+                    i++;
+                }
+
+                if( i - startOfFieldOrValue == 0U )
+                {
+                    /* A field should never be empty, but a value can be empty
+                     * provided a field was specified first. */
+                }
+                else if( !fieldHasValue )
+                {
+                    setQueryParameterKey( currentParameter, &pQuery[ startOfFieldOrValue ], i - startOfFieldOrValue, pCanonicalRequest );
+                    /* The previous field did not have a value set for it, so set its value to NULL. */
+                    setQueryParameterValue( currentParameter, NULL, 0U, pCanonicalRequest );
+                    startOfFieldOrValue = i + 1U;
+                    currentParameter++;
+                }
+                else
+                {
+                    /* End of value reached, so store a pointer to the previously set value. */
+                    setQueryParameterValue( currentParameter, &pQuery[ startOfFieldOrValue ], i - startOfFieldOrValue, pCanonicalRequest );
+                    fieldHasValue = 0U;
+                    startOfFieldOrValue = i + 1U;
+                    currentParameter++;
+                }
+            }
+            else
+            {
+                /* Empty else. */
+            }
+
+            if( currentParameter > SIGV4_MAX_QUERY_PAIR_COUNT )
```

Comment:
Logging an error will be helpful here . Also, I think this function should return Sigv4Status currently its returning void.

Reply:
Added and updated return type